### PR TITLE
Various tidyup for container `cli_wrapper.py`

### PR DESCRIPTION
**Pull request by LewisGaul**
- Correct some type annotations
- Move some formatting helpers to `utils.py` (to be used by the `pod` component too)
- More consistency in constructing commands, e.g. using helpers such as `full_cmd.add_simple_arg()`
- Move `exists()` method to expected place in the classes

### CODE REVIEW

File: python_on_whales/components/container/cli_wrapper.py [hunk 1]
```diff
@@ -563,31 +560,31 @@ def create(
...
         label_files: List[ValidPath] = [],
         link: List[ValidContainer] = [],
         link_local_ip: List[str] = [],
         log_driver: Optional[str] = None,
         log_options: List[str] = [],
         mac_address: Optional[str] = None,
-        memory: Union[int, str, None] = None,
-        memory_reservation: Union[int, str, None] = None,
-        memory_swap: Union[int, str, None] = None,
+        memory: Optional[Union[int, str]] = None,
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/container/cli_wrapper.py:585]: I'm not going to block this pull request for this, but note that this is not very useful to change, next python's release, we'll have `int | str | None` instead.

File: python_on_whales/components/container/cli_wrapper.py [hunk 2]
```diff
@@ -1829,14 +1818,15 @@ def unpause(self, x: Union[ValidContainer, List[ValidContainer]], /):
...
         """
-        x = to_list(x)
-        if x == []:
+        containers = to_list(x)
+        if len(containers) == 0:
             # nothing to do
             return
         full_cmd = self.docker_cmd + ["container", "unpause"]
-        full_cmd += x
+        full_cmd.extend(str(c) for c in containers)
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/container/cli_wrapper.py:1829]: To avoid forgetting the conversion to `str` every time we add something to `full_cmd`, maybe we should make `full_cmd` being a `list[Stringable]` with Stringable being a typing.Protocol that ensures that a method `__str__` is present. Then we can do `str()` when we call the function `run()`. What do you think? That would remove some complexity in the implementation of commands.

**Review by gabrieldemarmiesse — CHANGES_REQUESTED** [review]
very nice cleanup :)

File: python_on_whales/components/container/cli_wrapper.py (hunk 1 above)
Comment (LewisGaul) [reply python_on_whales/components/container/cli_wrapper.py:585]: Yeah sorry this was kind of a style preference change, I personally prefer using `Optional` to indicate arguments where `None` is used to represent "not specified" as opposed to having some other meaning (such as "no timeout"). It's then consistent with the other args that are `Optional[<something>]`, and doesn't look like the `| None` was missed.

I can change these back if you don't like this.

File: python_on_whales/components/container/cli_wrapper.py (hunk 2 above)
Comment (LewisGaul) [reply python_on_whales/components/container/cli_wrapper.py:1829]: Yeah, after making some of these changes I realised that `run()` already handles converting all args to strings, so this isn't strictly necessary... One option would be to reflect that in the type signature, although it does have the slight downside of accidentally passing in an object that doesn't have a suitable `__str__()` and it not being obvious when it doesn't work properly. Note that all objects are "stringable" in python.

File: python_on_whales/components/container/cli_wrapper.py (hunk 1 above)
Comment (gabrieldemarmiesse) [reply python_on_whales/components/container/cli_wrapper.py:585]: No worries, I won't block the PR for this, I'm just saying that it's not useful as it's very temporary.

File: python_on_whales/components/container/cli_wrapper.py (hunk 2 above)
Comment (gabrieldemarmiesse) [reply python_on_whales/components/container/cli_wrapper.py:1829]: Yeah haha, but it's the same issue here by calling `str()` manually, we can make the mistake and call it on the wrong objects.

File: python_on_whales/components/container/cli_wrapper.py (hunk 1 above)
Comment (LewisGaul) [reply python_on_whales/components/container/cli_wrapper.py:585]: Part of what I was saying is that even with the new syntax available I prefer sticking to `Optional` here, due to how it reads. That is, I'm agreeing with [this perspective](https://discuss.python.org/t/clarification-for-pep-604-is-foo-int-none-to-replace-all-use-of-foo-optional-int/26945/10).

File: python_on_whales/components/container/cli_wrapper.py (hunk 2 above)
Comment (LewisGaul) [reply python_on_whales/components/container/cli_wrapper.py:1829]: Well, I guess it's a bit more explicit when done here than in `run()`. But changing `run()` to *not* do it would presumably be difficult at this point. I started making these changes based on it being done in some methods and not others, before I realised `run()` does it automatically. Are you saying you want me to revert these stringify changes?

Comment (gabrieldemarmiesse) [reply python_on_whales/components/container/cli_wrapper.py:1829]: `run()` already does convert to string, so for me the changes you made are not needed. You can revert back to how it was.

File: python_on_whales/components/container/cli_wrapper.py (hunk 1 above)
Comment (gabrieldemarmiesse) [reply python_on_whales/components/container/cli_wrapper.py:585]: We can't let someone opinion be a standard for an implied meaning, notably because not enough people know about it so it's not going to be applied throughout the codebase, so the meaning in the codebase of `Optional[x]` won't be reliable. Even if the convention comes from a CPython core dev or other. 

As long as it's not in the python documentation or not in a PEP or not in a linter tool to enforce it, we can only assume true what is written in the python docs, that being that `Optional[x] == x | None`. Since the second option is shorter and doesn't require an import, it's going to be preferred in all cases thoughout the project, especially since we can enforce it automatically. We'll likely enforce it with pyupgrade or something similar because memorizing conventions when doing code reviews is a pain.

Long story short, if we adopt the convention proposed here, we have no way of enforce it thoughout the project, unless we force reviewers to force it, and it's very annoying. More automation is always preferable, we'll use pyupgrade for this.

File: python_on_whales/components/container/cli_wrapper.py [hunk 3]
```diff
@@ -795,26 +791,10 @@ def create(
 
         full_cmd.add_simple_arg("--workdir", workdir)
 
-        full_cmd.append(image)
-        full_cmd += command
+        full_cmd.append(str(image))
+        full_cmd.extend(command)
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/container/cli_wrapper.py:795]:
```suggestion
        full_cmd.append(image)
        full_cmd += command
```

File: python_on_whales/components/container/cli_wrapper.py [hunk 4]
```diff
@@ -1021,13 +1020,13 @@ def kill(
         full_cmd = self.docker_cmd + ["container", "kill"]
 
         full_cmd.add_simple_arg("--signal", format_signal_arg(signal))
-        full_cmd += containers
+        full_cmd.extend(containers)
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/container/cli_wrapper.py:1023]:
```suggestion
        full_cmd += containers
```
We can't accept diff that changes code style, unless the reason is very strong, otherwise, we'd get PRs to change code style all the time. The first person to write the line chooses the style.

File: python_on_whales/components/container/cli_wrapper.py [hunk 5]
```diff
@@ -1795,21 +1788,16 @@ def stop(
             time: Seconds to wait for stop before killing a container (default 10)
 
         # Raises
-            `python_on_whales.exceptions.NoSuchContainer` if the container does not exists.
+            `python_on_whales.exceptions.NoSuchContainer` if any of the
+            containers do not exist.
         """
         containers = to_list(containers)
-        if containers == []:
+        if len(containers) == 0:
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/container/cli_wrapper.py:1795]:
```suggestion
        if containers == []:
```
Same here, we don't get enough benefits to justify a small change like this. Otherwise we'd get pull requests with this kind of diff all the time.

**Review by gabrieldemarmiesse — CHANGES_REQUESTED** [review]

**Review by gabrieldemarmiesse — APPROVED** [review]
Thanks for the PR, looks good!
